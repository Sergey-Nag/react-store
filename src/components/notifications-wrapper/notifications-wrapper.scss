.notifications-wrapper {
  padding: 1rem;
  position: absolute;
  left: 0;
  top: 0;
  z-index: 100;
  pointer-events: none;
  display: flex;
  flex-direction: column-reverse;
  counter-set: alert;
}