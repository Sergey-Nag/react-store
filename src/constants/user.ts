export const USER_LOCAL_STATE_KEY = 'user-data';