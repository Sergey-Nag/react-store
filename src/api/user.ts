import { BOOKS_API_BASE_URL } from "../constants/books-api";
import { User } from "../types/user";
import { request } from "./reqest";

export const loginUser = async (username: string | null): Promise<User> => 
  await request<User>(`${BOOKS_API_BASE_URL}/signin`, 'POST', { username });
