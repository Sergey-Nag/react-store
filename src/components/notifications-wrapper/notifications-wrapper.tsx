import { NOTIFICATION_WRAPPER_ID } from "../../constants/notification-wrapper-id";
import './notifications-wrapper.scss';

export const NotificationsWrapper = () =>
  <div id={NOTIFICATION_WRAPPER_ID} className="notifications-wrapper"></div>;
