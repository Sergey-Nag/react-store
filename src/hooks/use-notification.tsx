import { createContext, useContext, useEffect, useState, ReactNode } from "react";
import { createPortal } from "react-dom";
import { Notification } from "../components/notification/notification";
import { NOTIFICATION_LIFETIME } from "../constants/notification-lifetime";
import { NOTIFICATION_WRAPPER_ID } from "../constants/notification-wrapper-id";
import { NotificationStyle } from "../enums/notification-style";

type NotificationContextType = {
  notify: (content: JSX.Element, type?: NotificationStyle) => void,
}

const initialState: NotificationContextType = {
  notify() {}
};

interface NotificationItem {
  content: JSX.Element,
  type: NotificationStyle
}

const NotificationContext = createContext(initialState);

export function NotificationsProvider({ children }: { children: ReactNode }) {
  const [notifications, setNotifications] = useState<NotificationItem[]>([]);

  const notify = (content: JSX.Element, type: NotificationStyle = NotificationStyle.default) => {
    setNotifications([...notifications, {content, type}]);
  }

  const renderNotifications = () => {
    const notificationElements = 
      notifications.map(({content, type}, i) => 
        <Notification type={type} key={`${i}-${type}`}>{content}</Notification>
      );

    return !!notifications.length && createPortal(
        notificationElements,
        document.getElementById(NOTIFICATION_WRAPPER_ID) as HTMLElement
      );
  }

  useEffect(() => {
    let timerId: ReturnType<typeof setTimeout>;

    if (notifications.length) {
      timerId = setTimeout(() => {
        setNotifications(([, ...restNotifications]) => [...restNotifications]);
      }, NOTIFICATION_LIFETIME);
    }

    return () => {
      clearTimeout(timerId);
    }
  }, [notifications]);

  return (
    <NotificationContext.Provider value={{ notify }}>
      { children }
      { renderNotifications() }
    </NotificationContext.Provider>
  );
}

export const useNotification = () => useContext(NotificationContext);
