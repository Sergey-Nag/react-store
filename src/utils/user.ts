import { USER_LOCAL_STATE_KEY } from "../constants/user";
import { User } from "../types/user";

export const getLocalStorageUser = (): User | null => {
  const user = localStorage.getItem(USER_LOCAL_STATE_KEY);

  return user ? JSON.parse(user) : null;
};

export const setLocalStorageUser = (user: User) =>
  localStorage.setItem(USER_LOCAL_STATE_KEY, JSON.stringify(user));

export const clearLocalStorageUser = () =>
  localStorage.removeItem(USER_LOCAL_STATE_KEY);
