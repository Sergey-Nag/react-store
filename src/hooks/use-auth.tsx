import { createContext, PropsWithChildren, useContext, useState } from "react"
import { loginUser } from "../api/user";
import { User } from "../types/user";
import { clearLocalStorageUser, getLocalStorageUser, setLocalStorageUser } from "../utils/user";

interface AuthState {
  user: User | null,
  login: (username: string) => Promise<User | null>,
  logout: () => void,
  authError: Error | null
}

const initialAuthState: AuthState = {
  user: null,
  authError: null,
  login() { return Promise.resolve(null) },
  logout() {},
};

const AuthContext = createContext(initialAuthState);

export function AuthProvider(props: PropsWithChildren) {
  const [userData, setUserData] = useState<User | null>(getLocalStorageUser);
  const [authError, setAuthError] = useState<Error | null>(null);

  const login = async (username: string) => {
    if (userData) return null;
    let userResponse;
    try {
      userResponse = await loginUser(username);
      setUserData(userResponse);
      setLocalStorageUser(userResponse);
      setAuthError(null);
    } catch(e: any) {
      setAuthError(e);
    }

    return userResponse ?? null;
  };

  const logout = () => {
    clearLocalStorageUser();
    setUserData(null);
  };

  const authContextValue: AuthState = { 
    user: userData,
    authError,
    login,
    logout,
  }

  return <AuthContext.Provider value={authContextValue} {...props} />
}

export const useAuth = () => useContext(AuthContext);