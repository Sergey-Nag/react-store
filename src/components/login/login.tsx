import { useEffect, useState } from "react";
import { useNavigate } from "react-router-dom";
import { EMAIL_REGEXP } from "../../constants/email-regexp";
import { useAuth } from "../../hooks/use-auth";
import { useDebounce } from "../../hooks/use-debounce";

export function LoginComponent() {
  const { login, user } = useAuth();
  const [username, setUsername] = useState<string | null>(null);
  const [isLoading, setIsLoading] = useState(false);
  const [isValid, setIsValid] = useState(false);
  const navigate = useNavigate();
  const debouncedUsername = useDebounce(username, 200);

  const handleUsernameChange = ({ target }: React.ChangeEvent<HTMLInputElement>) => {
    setUsername(target.value);
  };

  const handleLogin = async () => {
    if (!username) return;
    setIsLoading(true);

    await login(username);

    setIsLoading(false);
  };

  useEffect(() => {
    if (user) navigate('/catalog');
  }, [user]);

  useEffect(() => {
    setIsValid(EMAIL_REGEXP.test(debouncedUsername));
  }, [debouncedUsername]);

  return (
    <div className="container h-100">
      <div className="row justify-content-center align-items-center h-100">
        <div className="col-4">
          <div className="card p-3">
            <div className="form-group">
              <input 
                type="text" 
                onChange={handleUsernameChange}
                className="form-control"
                placeholder="Email"
              />
              <div className="d-grid pt-3">
                <button
                  onClick={handleLogin}
                  disabled={!isValid || isLoading}
                  className="btn btn-primary"
                >
                  { isLoading && <span className="spinner-border spinner-border-sm me-2" role="status" aria-hidden="true"></span> }
                  { isLoading ? 'Loading' : 'Submit' }
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  );
}