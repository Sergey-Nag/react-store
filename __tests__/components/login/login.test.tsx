import { fireEvent, render, screen, waitFor } from "@testing-library/react";
import { BrowserRouter, Route, Routes } from "react-router-dom";
import { LoginComponent } from "../../../src/components/login/login";
import { USER_LOCAL_STATE_KEY } from "../../../src/constants/user";
import { AuthProvider } from "../../../src/hooks/use-auth";
import { loginUser } from '../../../src/api/user';
import { NotificationsProvider } from "../../../src/hooks/use-notification";
import { NotificationsWrapper } from "../../../src/components/notifications-wrapper/notifications-wrapper";

jest.mock('../../../src/api/user', () => ({
  loginUser: jest.fn()
}));

describe('Login', () => {
  const mockUser = {
    avatar: 'test',
    username: 'test@test.t',
    token: 'test'
  };

  const renderWithAuthProviderAndRouter = () =>
    render(
      <AuthProvider>
        <BrowserRouter>
          <Routes>
            <Route index element={<LoginComponent />} />
            <Route path='/catalog' element={<div>Catalog</div>} />
          </Routes>
        </BrowserRouter>
      </AuthProvider>
    );

  const renderWithNotificationsAuthAndRouter = () => 
      render(
      <AuthProvider>
        <NotificationsProvider>
          <NotificationsWrapper />
            <BrowserRouter>
            <Routes>
              <Route index element={<LoginComponent />} />
              <Route path='/catalog' element={<div>Catalog</div>} />
            </Routes>
          </BrowserRouter>
        </NotificationsProvider>
      </AuthProvider>
      )

  afterEach(() => {
    localStorage.clear();
    jest.resetAllMocks();
  });

  it('Should render login component', () => {
    const { asFragment } = renderWithAuthProviderAndRouter();

    expect(asFragment()).toMatchSnapshot();
  });

  it('Should show error notification if request fails', async () => {
    (loginUser as jest.Mock).mockRejectedValue(new Error('Something went wrong'));

    renderWithNotificationsAuthAndRouter();

    fireEvent.change(
      screen.getByRole('textbox'), 
      { target : { value: 'test@test.tt' }}
    );
    
    const submitButton = screen.getByRole('button');

    await waitFor(() => expect(submitButton).toBeEnabled());

    fireEvent.click(submitButton);

    expect(submitButton).toBeDisabled();

    await waitFor(() => {
      expect(screen.getByRole('alert')).toBeInTheDocument();
    });

    expect(screen.getByText('Something went wrong')).toBeInTheDocument();
    expect(submitButton).toBeEnabled();
  });

  describe('The Submit button shold be desabled with values', () => {
    it.each(['', '123', 'asd@', 'asd@asd.'])('"%s"', (value) => {
      renderWithAuthProviderAndRouter();

      const usernameInput = screen.getByRole('textbox');

      fireEvent.change(usernameInput, { target : { value }});

      expect(screen.getByRole('button')).toBeDisabled();
    });
  });

  it('The Submit button should be enabled with correct email', async () => {
    renderWithAuthProviderAndRouter();

    const usernameInput = screen.getByRole('textbox');

    fireEvent.change(usernameInput, { target : { value: 'test@test.tt' }});
    
    await waitFor(() => expect(screen.getByRole('button')).toBeEnabled());
  });

  it(`After the submit button is pressed, it must be in the loading state during the api request.
  After success request should redirect to the catalog page`, async () => {
    (loginUser as jest.Mock).mockResolvedValue(mockUser);

    renderWithAuthProviderAndRouter();

    fireEvent.change(
      screen.getByRole('textbox'), 
      { target : { value: 'test@test.tt' }}
    );
    
    const submitButton = screen.getByRole('button');
    await waitFor(() => expect(submitButton).toBeEnabled());

    fireEvent.click(submitButton);

    expect(submitButton).toBeDisabled();

    await waitFor(async () => expect(await screen.findByText('Catalog')).toBeInTheDocument());
  });

  it('Should redirect to the catalog page immediately if the user exists in local storage', () => {
    localStorage.setItem(USER_LOCAL_STATE_KEY, JSON.stringify(mockUser));

    renderWithAuthProviderAndRouter();

    expect(screen.getByText('Catalog')).toBeInTheDocument();
  });
});