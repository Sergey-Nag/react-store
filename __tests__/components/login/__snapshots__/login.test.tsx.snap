// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Should render login component 1`] = `
<DocumentFragment>
  <div
    class="container h-100"
  >
    <div
      class="row justify-content-center align-items-center h-100"
    >
      <div
        class="col-4"
      >
        <div
          class="card p-3"
        >
          <div
            class="form-group"
          >
            <input
              class="form-control"
              placeholder="Email"
              type="text"
            />
            <div
              class="d-grid pt-3"
            >
              <button
                class="btn btn-primary"
                disabled=""
              >
                Submit
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
